examples/step-75/step-75.cc

```diff
@@ -0,0 +1,1575 @@
+        },
+        system_matrix,
+        system_matrix);
+
+      system_matrix.compress(VectorOperation::add);
+
+      const auto p = system_matrix.local_range();
+      for (auto i = p.first; i < p.second; i++)
+        if (system_matrix(i, i) == 0.0 && constraints.is_constrained(i))
+          system_matrix.add(i, i, 1);
+    }
+
+    dealii::MatrixFree<dim, number> matrix_free;
+
+    AffineConstraints<number> constraints;
+
+    std::vector<unsigned int> constrained_indices;
+
+    mutable std::vector<std::pair<number, number>> constrained_values;
+
+    mutable TrilinosWrappers::SparseMatrix system_matrix;
+
+    std::shared_ptr<const Utilities::MPI::Partitioner> partitioner_dealii;
+  };
+
+
+
+  // @sect3{The <code>Preconditioner</code> class template}
+
+  // Preconditioner for the equation system.
+  template <typename VectorType>
+  class PreconditionerBase
+  {
+  public:
+    virtual void vmult(VectorType &dst, const VectorType &src) const = 0;
+  };
+
+  template <typename VectorType>
+  class PreconditionerAMG : public PreconditionerBase<VectorType>
+  {
+  public:
+    template <typename Operator>
+    PreconditionerAMG(const Operator &system_matrix)
+    {
+      LA::MPI::PreconditionAMG::AdditionalData data;
+      data.elliptic              = true;
+      data.higher_order_elements = true;
+      preconditioner.initialize(system_matrix.get_system_matrix(), data);
+    }
+
+    void vmult(VectorType &dst, const VectorType &src) const override
+    {
+      preconditioner.vmult(dst, src);
+    }
+
+  private:
+    LA::MPI::PreconditionAMG preconditioner;
+  };
+
+
+
+  // @sect3{The <code>AdaptationStrategy</code> class template}
+
+  // Different strategies to perform hp adaptation.
+  template <int dim>
+  class AdaptationStrategy
+  {
+  public:
+    AdaptationStrategy(const ParameterHandler &                   params,
+                       DoFHandler<dim> &                          dof_handler,
+                       parallel::distributed::Triangulation<dim> &triangulation)
+      : params(params)
+      , dof_handler(dof_handler)
+      , triangulation(triangulation)
+    {
+      for (unsigned int degree = min_degree; degree <= max_degree; ++degree)
+        face_quadrature_collection.push_back(QGauss<dim - 1>(degree + 1));
+    };
+
+    virtual void
+    adapt_resolution(const LinearAlgebra::distributed::Vector<double>
+                       &locally_relevant_solution);
+
+  protected:
+    void flag_adaptation(const LinearAlgebra::distributed::Vector<double>
+                           &locally_relevant_solution);
+    virtual void decide_hp(const LinearAlgebra::distributed::Vector<double>
+                             &locally_relevant_solution);
+    void         limit_levels();
+    virtual void execute_refinement();
+
+    const ParameterHandler &params;
+    const unsigned int      min_level = 5, max_level = dim <= 2 ? 10 : 8;
+    const unsigned int      min_degree = 2, max_degree = dim <= 2 ? 7 : 5;
```

Comment:
we tend to not use initializers on const variables, and instead do this in the constructor

Reply:
This was just an intermediate construct :-)

I wanted to get these parameters from the `ParameterHandler` object. But I did not implement the parameter file part yet. This will be gone soon.